### PR TITLE
Queue entries outside of the event dispatching logic

### PR DESCRIPTION
**Pull request by npm1**
@tdresser @yoavweiss PTAL

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -318,7 +317,30 @@ When asked to <dfn>dispatch pending entries</dfn> for a {{Document}} <var>doc</v
                 1. Set <var>doc</var>'s <a>hasDispatchedEvent</a> to true.
                 1. <a>Queue the entry</a> <var>newFirstInputDelayEntry</var>.
 
-In the case where event handlers took no time, a user agent may opt not to queue the entry. This provides user agents with the flexibility to ignore input which never blocks on the main thread and for which event dispatching logic is skipped.
+Queueing other entries {#sec-queue-other}
+--------------------------------------------------------
+
+Some user agents skip the event dispatching logic for some events.
+In fact, this is theoretically possible for any event which does not have associated event handlers.
```
Comment (tdresser) [inline index.bs]: "associated event handlers or cause user agent defined behavior." (e.g., what used to be know as default actions)

File: index.bs [hunk 2]
```diff
@@ -318,7 +317,30 @@ When asked to <dfn>dispatch pending entries</dfn> for a {{Document}} <var>doc</v
...
+If the input is the first input, then the user agent must always queue the corresponding <code>"firstInput"</code> entry and it must do so via the event dispatching logic.
```
Comment (tdresser) [inline index.bs]: Should we be using some formalized MAY/MUST here? (I'm not sure, I've just seen folks do so)

File: index.bs [hunk 3]
```diff
@@ -318,7 +317,30 @@ When asked to <dfn>dispatch pending entries</dfn> for a {{Document}} <var>doc</v
...
+Otherwise, the user agent may opt to not queue the entry when there are no event handlers associated to the event.
```
Comment (tdresser) [inline index.bs]: "maybe opt not to" feels a bit clearer to me.

File: index.bs [hunk 4]
```diff
@@ -318,7 +317,30 @@ When asked to <dfn>dispatch pending entries</dfn> for a {{Document}} <var>doc</v
...
+This provides user agents with the flexibility to ignore input which never blocks on the main thread and for which event dispatching logic is skipped.
+
+To <em>create an event timing entry</em> outside of the event dispatching logic, the user agent must run the following steps:
+
+1. Let <var>event</var> be the event being measured and let <var>target</var> be the target {{Document}}.
+1. Run the <a>initialize event timing</a> algorithm, passing in <var>event</var> and replacing the <a>processing start step</a> with the following step:
```
Comment (tdresser) [inline index.bs]: This is pretty hard to follow.

Should we give the initialize/finalize event timing algorithms some additional parameters, to avoid the step replacement?

File: index.bs [hunk 5]
```diff
@@ -318,7 +317,30 @@ When asked to <dfn>dispatch pending entries</dfn> for a {{Document}} <var>doc</v
...
+    1. Set <var>timingEntry</var>'s {{processingStart}} to its {{PerformanceEntry/startTime}}.
+1. Let <var>timingEntry</var> be the output from the algorithm.
+1. Run the <a>finalize event timing</a> algorithm, passing in <var>timingEntry</var> and <var>target</var> as inputs, and replacing the <a>processing end step</a> with the following step:
+    1. Set <var>timingEntry</var>'s {{processingEnd}} to its {{PerformanceEntry/startTime}}.
+
+Note: processing the first input entry via the event dispatching logic ensures that the first input's {{PerformanceEntry/duration}} is calculated accurately.
```
Comment (tdresser) [inline index.bs]: Should we make this non-normative? or is "Note:" sufficient to do this?

File: index.bs [hunk 6]
```diff
@@ -318,7 +317,30 @@ When asked to <dfn>dispatch pending entries</dfn> for a {{Document}} <var>doc</v
...
+If skipping first inputs without event handlers was allowed, or if they were processed outside of event dispatching logic, then unwanted skews could happen.
```
Comment (tdresser) [inline index.bs]: skews could happen -> statistical biases could be introduced

I think that's a bit clearer.

File: index.bs (hunk 1 above)
Comment (npm1) [reply index.bs]: Done

File: index.bs (hunk 2 above)
Comment (npm1) [reply index.bs]: AFAIK must is already 'formalized' even if it's not in caps.

File: index.bs (hunk 3 above)
Comment (npm1) [reply index.bs]: Done

File: index.bs (hunk 4 above)
Comment (npm1) [reply index.bs]: Done. But in that case, since startTime was being computed in initialize, I've changed the return values to 0.

File: index.bs (hunk 5 above)
Comment (npm1) [reply index.bs]: I think that having it in a note is sufficient. @yoavweiss could confirm for sure

File: index.bs (hunk 6 above)
Comment (npm1) [reply index.bs]: Done

File: index.bs (hunk 5 above)
Comment (yoavweiss) [reply index.bs]: Typically we used something like `<p class=note>` around non-normative parts

File: index.bs [hunk 7]
```diff
@@ -318,7 +317,28 @@ When asked to <dfn>dispatch pending entries</dfn> for a {{Document}} <var>doc</v
...
+Queueing other entries {#sec-queue-other}
+--------------------------------------------------------
+
+Some user agents skip the event dispatching logic for some events.
+In fact, this is theoretically possible for any event which does not have associated event handlers or cause user agent defined behavior.
+If the input is the first input, then the user agent must always queue the corresponding <code>"firstInput"</code> entry and it must do so via the event dispatching logic.
+Otherwise, the user agent may opt not to queue the entry when there are no event handlers associated to the event.
+This provides user agents with the flexibility to ignore input which never blocks on the main thread and for which event dispatching logic is skipped.
+
+To <em>create an event timing entry</em> outside of the event dispatching logic, the user agent must run the following steps:
```
Comment (yoavweiss) [inline index.bs]: Who is calling this algorithm and when?

File: index.bs [hunk 8]
```diff
@@ -318,7 +317,28 @@ When asked to <dfn>dispatch pending entries</dfn> for a {{Document}} <var>doc</v
...
+
+1. Let <var>event</var> be the event being measured and let <var>target</var> be the target {{Document}}.
+1. Run the <a>initialize event timing</a> algorithm, passing in <var>event</var> and <code>0</code> as inputs.
```
Comment (yoavweiss) [inline index.bs]: Not clear to me why the time here should be 0, but it's probably because I'm lacking the context.

File: index.bs (hunk 2 above)
Comment (yoavweiss) [reply index.bs]: It is, but caps help to make that clearer.

Comment (npm1) [reply index.bs]: Ok done

File: index.bs (hunk 5 above)
Comment (npm1) [reply index.bs]: Bikeshed does this by default with paragraphs starting with "Note:".

File: index.bs (hunk 7 above)
Comment (npm1) [reply index.bs]: This is called when there are no event handlers associated to the event, the event dispatching logic is skipped, and the user agent opts to queue the entry. I'd like to be able to point out a specific place where it would get called but not sure what that would be...

File: index.bs (hunk 8 above)
Comment (npm1) [reply index.bs]: There is no event processing in this case (event dispatching logic was skipped) so there's no value that makes sense. Passing 0 as the `processingEnd` helps developers identify these cases.

**Comment by npm1** [comment]
Addressed the feedback here. I should point out that I'll need to automate the bikeshed part so the bs/html files are sync'd automatically.

File: index.bs (hunk 5 above)
Comment (yoavweiss) [reply index.bs]: missed the fact that this is Bikeshed... Yeah, "Note:" is enough

File: index.bs (hunk 8 above)
Comment (yoavweiss) [reply index.bs]: OK, got it

File: index.bs (hunk 7 above)
Comment (yoavweiss) [reply index.bs]: OK, that makes things clearer :)

Comment (yoavweiss) [reply index.bs]: Might be worthwhile to add a link to "event dispatching logic" (which I'm guessing is in HTML). I don't have opinions on where it should be called from.

Comment (npm1) [reply index.bs]: I've linked to the event dispatch algorithm from DOM spec now to make it more concrete.

**Comment by npm1** [comment]
Good question, what would you recommend for tests about "MAY" behavior? Here, the creation of entries outside of the event dispatch algorithm is optional (except for the first input), and relies on skipping event dispatch.

**Comment by npm1** [comment]
In Chrome, there could be events that are handled only by the compositor but require a lot of work within the compositor to be handled. They could theoretically take long to be processed even without having event handlers. But the point is that since both options (reporting, not reporting) are fine, a test for this would not be able to assert much.
